### PR TITLE
Use confirmation date for purchase order

### LINKED ISSUE #4831: Move start of lead time from purchase.created_at to the date of purchase confirmation
Currently, we base the calculation of lead times on the creation of a purchase order.  As described by our management consultant, a purchase order isn't sent to the supplier until it is approved by upper management - potentially several days later.  If this is the case, we will artificially inflate our lead times.

The lead time (delai de livraison) is important to estimate the amount of time it takes a supplier to deliver a medication.  If we are increasing that be computing the amount of time it takes to transmit the purchase order through the administrative process of the hospital, we may be introducing more noise into our calculation.

I propose that we keep a log of when purchase orders have been "confirmed" by the administration and use that as the start date of our lead time calculation.

See https://github.com/IMA-WorldHealth/bhima/issues/4797 for more discussion.

### PR DESCRIPTION
**Pull request by lomamech**
- Use confirmation date instead purchase date created
- Refactor of the calculation of the order interval as well as the number
  taking into account only purchase orders confirmed, received in stock
  (partially, in full, or even in excess)
- Recalculate the order interval each time the status of a purchase order is updated

closes #4831

### CODE REVIEW

**Comment by jmcameron** [comment]
bors try

File: test/integration/helpers.js [hunk 1]
```diff
@@ -209,9 +209,11 @@ exports.data = {
   PRICE_LIST : '75E09694DD5C11E5A8A26C29955775B0',
   SUPERUSER : 1,
   OTHERUSER : 2,
+  QUININE_TEXT : 'Quinine Bichlorhydrate, sirop, 100mg base/5ml, 100ml, flacon, Unité',
```
Comment (jniles) [inline test/integration/helpers.js:212]: Nice 👍

File: server/controllers/finance/purchases.js [hunk 2]
```diff
@@ -541,3 +550,83 @@ function purchaseState(req, res, next) {
...
+function purchaceIntervalleSetting(items) {
+  const trans = db.transaction();
+
+  items.forEach((item) => {
+    const sql = `
+      SELECT ROUND(AVG(orderIntervall.diff),2) AS purchase_interval, (COUNT(orderIntervall.text)) AS num_purchase,
+      BUID(?) AS inventory_uuid
+      FROM (
+        SELECT aggr.refPurchase, aggr.text, aggr.date, aggr.precedent_date,
+          (TIMESTAMPDIFF(DAY, DATE(aggr.precedent_date), DATE(aggr.date)) *12/365.24) AS diff,
```
Comment (jniles) [inline server/controllers/finance/purchases.js:563]: This is cool 👍   Never heard of TIMESTAMPDIFF before.

File: server/controllers/finance/purchases.js [hunk 3]
```diff
@@ -541,3 +550,83 @@ function purchaseState(req, res, next) {
...
+        aggr.inventory_uuid
+        FROM (
+          SELECT purch.refPurchase, purch.text, DATE(purch.date) AS date,
+            @last_purchase AS precedent_date, purch.statusText, (@last_purchase := DATE(purch.date)) AS last_purchase,
+            BUID(purch.inventory_uuid) AS inventory_uuid
+          FROM (
+            SELECT map.text AS refPurchase, inv.text, p.date, sta.text AS statusText, inv.uuid AS inventory_uuid
+            FROM purchase AS p
+            JOIN document_map AS map ON map.uuid = p.uuid
+            JOIN purchase_status AS sta ON sta.id = p.status_id
```
Comment (jniles) [inline server/controllers/finance/purchases.js]: Do you really need the JOIN on `purchase_status` and `document_map`?  You don't use them in the aggregating function.  If you do not need them, go ahead and remove them.

File: server/controllers/finance/purchases.js [hunk 4]
```diff
@@ -197,48 +210,42 @@ function create(req, res, next) {
...
-          * and in the following we would calculate each time the average value Multiply by number of old orders minus
-          * one Add up by the period between the last purchase Order and the date of the current purchase order
-          * and divide the results by the number of old orders
+        /*
+          * For the calculation of the average order interval, the system takes into account
+          * that of confirmed purchase orders, received in stock or partially and even those
+          * that are received in excess quantity, the default order interval is zero, and the
+          * calculation is made in addition the sum of all the order intervals by the sum of
+          * the purchase orders minus one, the results are stored in the inventory table,
+          * in order to facilitate the calculation of the various indicators for stock management
```
Comment (jniles) [inline server/controllers/finance/purchases.js:240]: Nice 👍

**Review by jniles — COMMENTED** [review]
Just one comment about the use of more JOINs than necessary.  Otherwise, LGTM.  You can merge it if you feel those changes are not necessary or if you complete them.

**Review by jniles — APPROVED** [review]
LGTM.

bors r+

**Pull request merged**